presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java

```diff
@@ -0,0 +1,101 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.client;
+
+import com.google.auth.oauth2.AccessToken;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.common.base.Splitter;
+import okhttp3.Interceptor;
+import okhttp3.Request;
+import okhttp3.Response;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import static com.facebook.presto.client.GCSOAuthScope.DEVSTORAGE_READ_ONLY;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_EXTRA_CREDENTIAL;
+import static java.util.Objects.requireNonNull;
+
+public class GCSOAuthInterceptor
+        implements Interceptor
+{
+    public static final String GCS_CREDENTIALS_PATH_KEY = "hive.gcs.credentials.path";
+    public static final String GCS_OAUTH_SCOPES_KEY = "hive.gcs.oauth.scopes";
+
+    private static final String GCS_CREDENTIALS_OAUTH_TOKEN_KEY = "hive.gcs.oauth";
+    private static final Splitter SCOPE_SPLITTER = Splitter.on("|");
+
+    private final Collection<String> gcsOAuthScopeURLs;
+    private final String credentialsFilePath;
+
+    private GoogleCredentials credentials;
+
+    public GCSOAuthInterceptor(String credentialPath, Optional<String> gcsOAuthScopesString)
+    {
+        this.credentialsFilePath = requireNonNull(credentialPath);
+        this.gcsOAuthScopeURLs = mapScopeStringToURLs(requireNonNull(gcsOAuthScopesString).orElse(DEVSTORAGE_READ_ONLY.name()));
+    }
+
+    @Override
+    public Response intercept(Chain chain)
+            throws IOException
+    {
+        return chain.proceed(attachGCSAccessToken(chain.request()));
+    }
+
+    private Request attachGCSAccessToken(Request request)
+    {
+        AccessToken token = getCredentials().getAccessToken();
+        return request.newBuilder()
+                .addHeader(PRESTO_EXTRA_CREDENTIAL, GCS_CREDENTIALS_OAUTH_TOKEN_KEY + "=" + token.getTokenValue())
+                .build();
+    }
+
+    private synchronized GoogleCredentials getCredentials()
+    {
+        if (credentials == null) {
+            credentials = createCredentials();
+        }
+        try {
+            credentials.refreshIfExpired();
+        }
+        catch (IOException e) {
+            throw new ClientException("Google credential refreshing error", e);
+        }
+        return credentials;
+    }
+
+    private GoogleCredentials createCredentials()
+    {
+        try {
+            return GoogleCredentials.fromStream(new FileInputStream(credentialsFilePath)).createScoped(gcsOAuthScopeURLs);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *FileInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when GoogleCredentials.fromStream throws an exception. Ensure the object returned by *FileInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).